### PR TITLE
Add sharing support

### PR DESCRIPTION
**Pull request by Gara64**
Implements sharing capabilities for the data-system, as described [here](https://github.com/Ljinod/cozy-docs/blob/master/src/documents/en/hack/cookbooks/sharing.html.md).

The logic is split in 2 files : 

- controllers/sharing -> the sharing process, for the creation, deletion, answering, etc.
- lib/sharing -> sharing methods to replicate data, send notification, etc.

Note this code can only work if the proxy and home are able to receive and interpret sharing requests. 
But this will come later :)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,7 +27,8 @@
     "request": "2.67.0",
     "request-json": "0.5.5",
     "string": "1.8.0",
-    "through": "2.3.7"
+    "through": "2.3.7",
+    "async": "~1.5.2"
```
Comment (nono) [inline package.json]: We prefer to have exact dependencies (`1.5.2` instead of `~1.5.2`), as we had some issues in the past with not doing that.

File: server/controllers/sharing.coffee [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+        err = new Error "Bad request"
+        err.status = 400
+        return next err
+    
+    # The docType is fixed
+    share.docType = "sharing"
+
+    # Generate a preToken for each target
+    for target in share.targets
+        do (target) ->
```
Comment (nono) [inline server/controllers/sharing.coffee]: I think this line (`do (target) ->`) is useless here

File: server/lib/request.coffee [hunk 3]
```diff
@@ -239,6 +239,17 @@ initializeDSView = (callback) ->
                 }
                 """
 
+        # Useful to manage sharing
+        sharing:
+            all:
+                map: """
+                function(doc) {
+                    if(doc.docType && doc.docType.toLowerCase() === "sharing") {
+                        return emit(doc._id, doc);
```
Comment (nono) [inline server/lib/request.coffee]: It's a good practice to only emit only the _id, not the full document. You can retrieve the full document with `include_docs`.

File: server/lib/sharing.coffee [hunk 4]
```diff
@@ -0,0 +1,109 @@
+db = require('../helpers/db_connect_helper').db_connect()
+async = require 'async'
+request = require 'request-json'
+log = require('printit')
+    prefix: 'sharing'
+
+# Get the Cozy url
+getDomain = (callback) ->
```
Comment (nono) [inline server/lib/sharing.coffee]: Maybe we can share some code with https://github.com/cozy/cozy-data-system/blob/master/server/controllers/mails.coffee#L128-L137

File: server/lib/sharing.coffee [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    checkDomain params, (err, params) ->
+
+        remote = request.createClient params.url
+        remote.post path, params, (err, result, body) ->
+            if err?
+                callback err
+            else if not result?.statusCode?
+                err = new Error "Bad request"
+                err.status = 400
+                callback err
```
Comment (nono) [inline server/lib/sharing.coffee]: What if `result.statusCode` is 5xx? We should probably consider it an error even if the body has no `error` key.

File: server/lib/sharing.coffee [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    unless replicationID?
+        err = new Error 'Parameters missing'
+        err.status = 400
+        callback err
+    else
+        cancel =
+            replication_id: replicationID
+            cancel: true
+
+        db.replicate '', cancel, (err, body) ->
```
Comment (nono) [inline server/lib/sharing.coffee]: Can you explain me what is the first parameter of db.replicate (an empty string here)?

**Comment by nono** [comment]
Can you add some unit tests?

File: server/lib/request.coffee (hunk 3 above)
Comment (Gara64) [inline server/lib/request.coffee]: Agreed, I'll change this

File: server/lib/sharing.coffee (hunk 6 above)
Comment (Gara64) [inline server/lib/sharing.coffee]: This is the cradle syntax to replicate (or cancel) on the database level : https://github.com/flatiron/cradle#database-level. 
In this case the first parameter is supposed to be the url, but it is useless as the replication_id is enough. I tried with the url (worked) and without (didn't work as it considers on the server level).

**Comment by Gara64** [comment]
@nono, yep, @Ljinod is on it :)

File: server/lib/sharing.coffee (hunk 6 above)
Comment (nono) [inline server/lib/sharing.coffee]: OK. Thanks for the info!

File: package.json (hunk 1 above)
Comment (Gara64) [inline package.json]: Ok didn't know that, I'll put `1.5.2`

File: server/lib/sharing.coffee (hunk 5 above)
Comment (Gara64) [inline server/lib/sharing.coffee]: You're right, I'll add it

**Comment by frankrousseau** [comment]
@Gara64 @nono Can we merge that one?

**Comment by nono** [comment]
@frankrousseau nope, it's not ready

@Gara64 can you look at linting? Travis builds show some errors with it

@Ljinod do you need some help for the unit tests?

**Comment by Ljinod** [comment]
@nono I'm okay for now and I should have the tests for `controllers/sharing` soon

**Comment by Gara64** [comment]
@nono I fixed the linting and added @Ljinod's tests units. I still need to fix 3 tests that failed though.

**Comment by nono** [comment]
@Gara64 Good news!

**Comment by Gara64** [comment]
@nono Seems ok now :)

File: package.json [hunk 7]
```diff
@@ -28,7 +27,8 @@
     "request": "2.67.0",
     "request-json": "0.5.5",
     "string": "1.8.0",
-    "through": "2.3.7"
+    "through": "2.3.7",
+    "async": "1.5.2"
```
Comment (nono) [inline package.json]: We try to keep dependencies in alphabetical order. It's what `npm install --save` does and we do the same to minimize diff (easier to review).

File: package.json [hunk 8]
```diff
@@ -37,9 +37,11 @@
     "coffee-script": "1.10.0",
     "coffeelint": "1.13.0",
     "mocha": "1.17.1",
-    "should": "3.1.2"
+    "should": "3.1.2",
+    "lodash": "4.6.1",
+    "sinon": "1.17.3"
   },
-  "main": "build/server.js",
+  "main": "server.coffee",
```
Comment (nono) [inline package.json]: You should keep it as it was

File: server/controllers/replication.coffee [hunk 9]
```diff
@@ -132,12 +132,13 @@ module.exports.proxy = (req, res, next) ->
...
-                        # Check document docType
+                        # Check document docType and id
                         if doc
-                            err = checkPermissions req, doc.docType
+                            docInfo = {id: doc._id, docType: doc.docType}
+                            err = checkPermissions req, docInfo
                             if err
                                 # Device isn't authorized
-                                res.status(403).send err
+                                res.send 403, err
```
Comment (nono) [inline server/controllers/replication.coffee]: Why this change? I think that new versions of express have deprecated res.send with a status code

File: server/controllers/sharing.coffee [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+            url: target.url
+            token: if target.token? then target.token else target.preToken
+            shareID: share.shareID
+            desc: "The sharing #{share.shareID} has been deleted"
+
+        log.info "Send sharing cancel notification to : #{notif.url}"
+
+        Sharing.notifyTarget "services/sharing/cancel", notif,
+        (err, result) ->
+            callback err
```
Comment (nono) [inline server/controllers/sharing.coffee]: Why not just this?

```coffee
Sharing.notifyTarget "services/sharing/cancel", notif, callback
```

File: server/controllers/sharing.coffee [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+#   targets[]  -> the targets to notify. Each target must have an url
+#                 and a token
+module.exports.sendDeleteNotifications = (req, res, next) ->
+    share = req.share
+
+    # Notify each target
+    async.each share.targets, (target, callback) ->
+        notif =
+            url: target.url
+            token: if target.token? then target.token else target.preToken
```
Comment (nono) [inline server/controllers/sharing.coffee]: You can make it simpler:

```coffee
token: target.token or target.preToken
```

File: server/lib/request.coffee (hunk 3 above)
Comment (nono) [inline server/lib/request.coffee]: It is still the case

File: package.json (hunk 7 above)
Comment (Gara64) [inline package.json]: Ok, good to know

File: package.json (hunk 8 above)
Comment (Gara64) [inline package.json]: Yes, fixed

File: server/controllers/replication.coffee (hunk 9 above)
Comment (Gara64) [inline server/controllers/replication.coffee]: I'm not sure what happened here. Anyway, it's fixed

**Comment by nono** [comment]
@frankrousseau do you want to review this PR too?

**Comment by aenario** [comment]
A few remarks : 

- There will propably be some conflicts with master on build & package.json. It might be more efficient to rebase / merge master before putting it into sharing. 

- Nitpicking but in controllers/sharing.coffee#73 : any reason not to juste do
`unless rule.docType and rule.id`

- In controllers/replication, why creating the object `docInfo` instead of passing directly bodyToTransmit ? It will be 1 - less RAM-hungry (by some byte :neutral_face: ) 2 - future-proof if we want to have sharing by tags or other fun things later. 

- The style of sharing tests is not the same than for other controllers : 
we usually test the HTTP interface
whereas these test the controllers/sharing functions unitly. 

The unit approach makes for great speed, but the http approach ensure everything works (for instance here, you dont test the correct ID is passed to db.delete, only that db.delete has been called).

Not sure which approach is best, may be we neeed both, but it's the reason you have to stub a lot of things with all the XXX Ugly parts (which could be replaced by https://github.com/howardabrams/node-mocks-http#usage)

**Comment by frankrousseau** [comment]
Yep, I would like to review this one before merging.

**Comment by Gara64** [comment]
@aenario About your three first remarks : I agree and I changed accordingly.
I let @Ljinod answer for the unit tests.

File: server/controllers/routes.coffee [hunk 12]
```diff
@@ -328,8 +329,32 @@ module.exports =
...
         ]
+    'services/sharing/sendAnswer':
+        post: [
+            utils.checkPermissionsFactory 'sharing'
+            sharing.handleRecipientAnswer
+            sharing.sendAnswer
+        ]
+    'services/sharing/answer':
+        post: [
+            utils.checkPermissionsFactory 'sharing'
```
Comment (frankrousseau) [inline server/controllers/routes.coffee]: Why do you use a factory based on function names?

Comment (Gara64) [inline server/controllers/routes.coffee]: The goal is to use the sharing as a service, so we can't base the permissions on the document level.
Ideally, we would need a more fine-grained permission system, because an application with sharing permission shouldn't have full access to this service.
But by the time we have this kind of permission system, I think this way is good enough.

Comment (frankrousseau) [inline server/controllers/routes.coffee]: I was thinking about a single function:

    utils.checkPermissionsSharing

Comment (Gara64) [inline server/controllers/routes.coffee]: Ah ok. As it would do the same job (redirection to checkPermissions), it didn't seem useful to me. Is it a matter of style?

Comment (frankrousseau) [inline server/controllers/routes.coffee]: Yes doing a factory for generating two functions look useless.

Comment (Gara64) [inline server/controllers/routes.coffee]: The factory is already used for other purposes, eg access, so isn't it better to avoid to create a new permission function?

File: server/controllers/sharing.coffee [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+    unless share.rules?.length > 0
+        err = new Error "No rules specified"
+        err.status = 400
+        return next err
+    # ...and well formed
+    for rule in share.rules
+        unless rule.docType and rule.id
+            err = new Error "Incorrect rule detected"
+            err.status = 400
+            return next err
```
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: Breaking a for loop with a return is ugly.

File: server/controllers/sharing.coffee [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+#   rules[]    -> the set of rules specifying which documents are shared,
+#                 with their docTypes.
+#   targets[]  -> the targets to notify. Each target must have an url
+#                 and a preToken
+module.exports.sendSharingRequests = (req, res, next) ->
+
+    share = req.share
+
+    # Notify each target
+    async.each share.targets, (target, callback) ->
```
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: Prefer `eachSeries` over `series`, it leads to less side effects.

File: server/controllers/sharing.coffee [hunk 15]
```diff
@@ -0,0 +1,434 @@
...
+#                This will be used as the sharer's login
+#   accepted  -> boolean specifying if the share was accepted or not
+#   preToken  -> the token sent by the sharer to authenticate the receiver
+#   url       -> the url of the cozy
+#   hostUrl   -> the url of the sharer's cozy
+#   rules     -> the set of rules specifying which documents are shared,
+#                with their docTypes.
+module.exports.handleRecipientAnswer = (req, res, next) ->
+
+    if not req.body? or Object.keys(req.body).length == 0
```
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: `is` instead of `==`

File: server/controllers/sharing.coffee [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+    share = req.body
+
+    # Check share structure: each element must not be null/undefined/empty
+    if not share.id?    or share.id is ''          or
+    not share.shareID?  or share.shareID is ''     or
+    not share.preToken? or share.preToken is ''    or
+    not share.accepted? or share.accepted is ''    or
+    not share.url?      or share.url is ''         or
+    not share.hostUrl?  or share.hostUrl is ''     or
+    not share.rules?    or not share.rules.length? or share.rules.length == 0
```
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: This condition could probably be built via a function like


```coffeescript
isEmptyFields = (obj) ->
    return not _.reduce(object.keys(obj), (sum, key) -> 
          value = obj[key]
          return sum and (value? and value isnt '' and value.length isnt 0) 

cond1 = not isEmptyFields share
```
(not tested)

File: server/controllers/sharing.coffee [hunk 17]
```diff
@@ -0,0 +1,434 @@
...
+    # Same for the rules structure: each rule must have an id and a docType
+    for rule in share.rules
+        if not rule.id?   or rule.id is ''      or
+        not rule.docType? or rule.docType is ''
+            err = new Error "Bad request: incorrect rule detected"
+            err.status = 400
+            return next err
+
+    # Create an access if the sharing is accepted
+    if share.accepted is yes
```
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: We prefer `true` over `yes`

File: server/controllers/sharing.coffee (hunk 16 above)
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: Sorry the example above sucks, it should rely on a while loop to avoid useless operation.

File: server/controllers/sharing.coffee [hunk 18]
```diff
@@ -0,0 +1,434 @@
...
+            rules: share.rules
+
+        libToken.addAccess access, (err, doc) ->
+            return next err if err?
+
+            share.token = access.password
+            req.share = share
+            return next()
+
+        # TODO : enforce the docType protection with the couchDB's document
```
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: Can you do it now?

File: server/controllers/sharing.coffee [hunk 19]
```diff
@@ -0,0 +1,434 @@
...
+        err = new Error "Bad request: body is incomplete"
+        err.status = 400
+        return next err
+
+    # Get the Sharing document thanks to its id
+    db.get answer.shareID, (err, doc) ->
+        return next err if err?
+
+        # Get the answering target
+        target = (t for t in doc.targets when t.url is answer.hostUrl)
```
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: maybe that `doc.targets.filter (target) -> url is answer.hostUrl` is more explicit for that case.

File: server/controllers/sharing.coffee [hunk 20]
```diff
@@ -0,0 +1,434 @@
...
+                next err
+            else
+                # Update the target with the repID if the sharing is continuous
+                if replicate.continuous
+                    db.get replicate.id, (err, doc) ->
+                        return next err if err?
+
+                        # Find the target in the db
+                        targetUrl = replicate.target.url
+                        target = (t for t in doc.targets when \
```
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: Maybe you should write an helper function `getTargets`.

File: server/controllers/sharing.coffee [hunk 21]
```diff
@@ -0,0 +1,434 @@
...
+        res.status(200).send success: true
+
+# Stop current replications for each specified target
+# Params must contain:
+#   targets[]  -> Each target must have an url,  a repID and a token
+module.exports.stopReplications = (req, res, next) ->
+    share = req.share
+
+    # Cancel the replication for all the targets
+    async.each share.targets, (target, cb) ->
```
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: Please use `eachSeries`.

File: server/helpers/utils.coffee [hunk 22]
```diff
@@ -24,20 +26,73 @@ module.exports.checkPermissions = (req, permission, next) ->
...
-        err = new Error "Application is not authorized"
+        err = new Error "Device #{login} is not authorized"
         err.status = 403
         return err
+    # Permissions are correct
     else
-        feed.publish 'usage.application', appName
-        req.appName = appName
+        feed.publish 'usage.application', login
+        req.login = login
```
Comment (frankrousseau) [inline server/helpers/utils.coffee]: Are you sure it doesn't break things?

File: server/lib/indexer.coffee [hunk 23]
```diff
@@ -25,40 +27,47 @@ indexdefinitionsID = {}
...
+
 ###*
 # Initialize the indexer
 #
 # @return (callback) when initialization is complete
 ###
 exports.initialize = (callback) ->
 
-    async.waterfall [
+    async.series [
```
Comment (frankrousseau) [inline server/lib/indexer.coffee]: Subfunctions are better than waterfall/series calls.

File: server/lib/sharing.coffee [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+#   docIDs     -> the ids of the documents to replicate
+#   continuous -> [optionnal] if the sharing is synchronous or not
+module.exports.replicateDocs = (params, callback) ->
+    unless params.target? and params.docIDs? and params.id?
+        err = new Error 'Parameters missing'
+        err.status = 400
+        callback err
+    else
+        # Add the credentials in the url
+        auth = params.id + ":" + params.target.token
```
Comment (frankrousseau) [inline server/lib/sharing.coffee]: It's better to use coffeescript notation: "#{params.id}:#{params.target.token}".

File: server/lib/sharing.coffee [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+        url = params.target.url.replace "://", "://" + auth + "@"
```
Comment (frankrousseau) [inline server/lib/sharing.coffee]: It's better to use coffeescript notation for strings.

File: server/lib/sharing.coffee [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+
+        replication =
+            source: "cozy"
+            target: url + "/services/sharing/replication/"
+            continuous: params.continuous or false
+            doc_ids: params.docIDs
+
+        log.info  "Replicate " + JSON.stringify params.docIDs + " to " + url
```
Comment (frankrousseau) [inline server/lib/sharing.coffee]: You should not write in logs an url containing credentials.

File: server/lib/token.coffee [hunk 27]
```diff
@@ -4,39 +4,48 @@ log = require('printit')
...
 ## @auth {string} Field 'authorization' of request
-## @tokens {tab} Tab which contains applications and their tokens
+## @tokensTab {tab} Tab which contains applications and their tokens
 ## @callback {function} Continuation to pass control back to when complete.
 ## Check if application is well authenticated
-checkToken = module.exports.checkToken = (auth) ->
+checkToken = module.exports.checkToken = (auth, tokensTab) ->
     if auth isnt "undefined" and auth?
+        # Default case for tokens
+        unless tokensTab? then tokensTab = tokens
```
Comment (frankrousseau) [inline server/lib/token.coffee]: It's better to write:

```coffeescript
tokensTab = tokens unless tokensTab?
```

File: server/lib/token.coffee [hunk 28]
```diff
@@ -4,39 +4,48 @@ log = require('printit')
...
 ## @docType {String} document's docType that application want manage
-## @name {String} application's name
 ## @callback {function} Continuation to pass control back to when complete.
 ## Check if application can manage docType
 module.exports.checkDocType = (auth, docType, callback) ->
     # Check if application is authenticated
 
     if productionOrTest
-        [err, isAuthenticated, name] = checkToken auth
+        [err, isAuthenticated, name] = checkToken auth, tokens
```
Comment (frankrousseau) [inline server/lib/token.coffee]: This line is similar to both conditions. It could be written before the `if productionOrTest`.

File: server/lib/token.coffee [hunk 29]
```diff
@@ -81,10 +89,73 @@ module.exports.checkDocTypeSync = (auth, docType) ->
...
+## @auth {string} Field 'authorization' of request
+## @doc {Object} Contains the document to evaluate
+## @callback {function} Continuation to pass control back to when complete.
+## Check if the sharing can manage the document
+module.exports.checkSharingRule = (auth, doc, callback) ->
+    if productionOrTest
+        # Check if requester is authenticated
+        [err, isAuthenticated, name] = checkToken auth, sharingTokens
+        if isAuthenticated
+            if doc?.id? && doc?.docType?
```
Comment (frankrousseau) [inline server/lib/token.coffee]: prefer `and` over `&&`

File: server/lib/token.coffee [hunk 30]
```diff
@@ -81,10 +89,73 @@ module.exports.checkDocTypeSync = (auth, docType) ->
...
+                doc.docType = doc.docType.toLowerCase()
+                # Check all the sharing rules defined for this login
+                for rule in sharingPermissions[name]
+                    if rule.id is doc.id && rule.docType is doc.docType
```
Comment (frankrousseau) [inline server/lib/token.coffee]: prefer `and` over `&&`

File: server/lib/token.coffee [hunk 31]
```diff
@@ -81,10 +89,73 @@ module.exports.checkDocTypeSync = (auth, docType) ->
...
+                        # Request is granted
+                        return callback null, name, true
+                # Request is denied
+                callback null, name, false
```
Comment (frankrousseau) [inline server/lib/token.coffee]: This could have been written more elegantly with:

```coffeescript
rule = _.findWhere sharingPermissions[name], {id: doc.id, docType: docType}
callback null, name, rule?
```

File: server/lib/token.coffee [hunk 32]
```diff
@@ -81,10 +89,73 @@ module.exports.checkDocTypeSync = (auth, docType) ->
...
+## Check if the sharing can manage the document
+module.exports.checkSharingRuleSync = (auth, doc) ->
+    if productionOrTest
+        # Check if requester is authenticated
+        [err, isAuthenticated, name] = checkToken auth, sharingTokens
+        if isAuthenticated
+            if doc?.id? && doc?.docType?
+                doc.docType = doc.docType.toLowerCase()
+                # Check all the sharing rules defined for this login
+                for rule in sharingPermissions[name]
```
Comment (frankrousseau) [inline server/lib/token.coffee]: same here, at first sight, findWhere looks to do a better job.

File: server/lib/token.coffee [hunk 33]
```diff
@@ -81,10 +89,73 @@ module.exports.checkDocTypeSync = (auth, docType) ->
...
+                    if rule.id is doc.id && rule.docType is doc.docType
+                        # Request is granted
+                        return [null, name, true]
+                # Request is denied
+                return [null, name, false]
+            else
+                # Particular case for couchDB requests without id or docType
+                return [null, name, true]
```
Comment (frankrousseau) [inline server/lib/token.coffee]: Why requests without id or docType are authenticated?

**Comment by frankrousseau** [comment]
I made some comments about code styles. After that I vote for a merge of this PR.

File: server/lib/token.coffee (hunk 27 above)
Comment (nono) [inline server/lib/token.coffee]: Why not `tokensTab ?= token` ?

File: server/lib/token.coffee (hunk 33 above)
Comment (Gara64) [inline server/lib/token.coffee]: To allow the CouchDB's replication requests that don't have documents. When they do, we control the id and docType before. See [here](https://github.com/Gara64/cozy-data-system/blob/sharing_experimentations/server/middlewares/utils.coffee#L90)

File: server/controllers/sharing.coffee (hunk 18 above)
Comment (Gara64) [inline server/controllers/sharing.coffee]: As it is not completely crucial  and could be used in other contexts than the sharing, I'd prefer to keep it for later to do it properly.

File: server/helpers/utils.coffee (hunk 22 above)
Comment (Gara64) [inline server/helpers/utils.coffee]: It does not break the unit tests, but I redo `req.appName` just to be sure

File: server/controllers/sharing.coffee (hunk 18 above)
Comment (frankrousseau) [inline server/controllers/sharing.coffee]: ok

File: server/controllers/routes.coffee (hunk 12 above)
Comment (frankrousseau) [inline server/controllers/routes.coffee]: ok
